### PR TITLE
Fix fd leak when connect() fails

### LINKED ISSUE #2994: Too many open files
After about 37 hours we encounter a "Too many open files" crash on several host on the cloud testnet.

startime: 2019-07-24 03:00:25.476148Z
crashtime: 2019-07-25 16:18:57.397740Z

error:

```
    (monitor.ml.Error (Unix.Unix_error "Too many open files" accept "((fd 43))")
  ("Raised at file \"src/core_unix.ml\", line 50, characters 4-43"
    "Called from file \"src/core_unix.ml\", line 2419, characters 17-40"
    "Called from file \"src/raw_fd.ml\", line 282, characters 10-26"
    "Re-raised at file \"src/unix_syscalls.ml\", line 956, characters 20-29"
    "Called from file \"src/unix_syscalls.ml\", line 961, characters 12-32"
    "Called from file \"src/deferred1.ml\", line 174, characters 6-13"
    "Called from file \"src/deferred0.ml\", line 49, characters 2-10"
    "Called from file \"src/unix_syscalls.ml\", line 983, characters 14-47"
    "Called from file \"src/unix_syscalls.ml\", line 1005, characters 6-78"
    "Called from file \"src/tcp.ml\", line 357, characters 6-89"
    "Called from file \"src/job_queue.ml\" (inlined), line 131, characters 2-5"
    "Called from file \"src/job_queue.ml\", line 170, characters 6-47"
    "Caught by monitor coda"))
```

### PR DESCRIPTION
**Pull request by enolan**
We were leaking the socket when connect() failed, since our logic for closing
them uses Reader.close, and no reader is created until after we have a
connection. We had our own code to set up sockets bound to a particular local
IP for outbound connections, but it turns out you can create a
Where_to_listen.t that includes the local IP. I changed it to do that, and to
pass the Where_to_listen.t to Rpc.Connection.with_client, which handles setting
up the sockets and disposing of the resources properly for us.

We were also calling Reader.close and Writer.close manually when handling error
conditions. This is redundant inside of a Tcp.Server.create callback, the
socket is closed automatically when the Deferred.t returned by the callback is
determined. Removed those calls.

Checklist:

- [ ] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [x] Does this close issues? List them:

Closes #2994

### CODE REVIEW

**Comment by enolan** [comment]
oops, bad merge

**Comment by bkase** [comment]
Do ya'll think this one is safe to land into release branch? Would really like to have the next release not have the file-descriptor problem we had last time.

File: src/lib/gossip_net/gossip_net.ml [hunk 1]
```diff
@@ -313,22 +270,9 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
   let try_call_rpc t (peer : Peer.t) dispatch query =
     (* use error collection, close connection strategy of Tcp.with_connection *)
```
Comment (cmr) [inline src/lib/gossip_net/gossip_net.ml]: can you remove this comment? this code no longer resembles with_connection.

File: src/lib/gossip_net/gossip_net.ml [hunk 2]
```diff
@@ -238,41 +237,6 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
...
-       but it only works on Linux. I think we don't need to worry about it,
-       there are a lot of ephemeral ports and we don't make very many
-       simultaneous connections.
-      *)
-      Async.Socket.bind_inet unbound_socket
-      @@ `Inet (t.addrs_and_ports.bind_ip, 0)
-    in
-    let peer_addr = `Inet (peer.Peer.host, peer.communication_port) in
-    try_with (fun () ->
-        let%map connected_socket = Async.Socket.connect socket peer_addr in
```
Comment (cmr) [inline src/lib/gossip_net/gossip_net.ml]: I guess this would throw, leading to the leak?

**Comment by cmr** [comment]
@bkase yes

File: src/lib/gossip_net/gossip_net.ml (hunk 1 above)
Comment (enolan) [reply src/lib/gossip_net/gossip_net.ml]: :+1:

File: src/lib/gossip_net/gossip_net.ml (hunk 2 above)
Comment (enolan) [reply src/lib/gossip_net/gossip_net.ml]: yeah.

**Pull request merged**